File: products/cpp/tf_pratice_ws/src/demo_cpp_tf2/src/static_tf_broadcaster.cpp
#include<memory>
#include "rclcpp/rclcpp.hpp"
#include"tf2_ros/static_transform_broadcaster.h"
#include"geometry_msgs/msg/transform_stamped.hpp"
#include"tf2/LinearMath/Quaternion.h"
#include <math.h>
using namespace std::chrono_literals;

class StaticTFBroadcaster:public rclcpp::Node
{
    public:
        StaticTFBroadcaster():Node("static_tf_broadcaster")
        {
            RCLCPP_INFO(this->get_logger(),"静态坐标转换发布节点创建");
            broadcaster_ = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
            publish_static_tf();
        }
    private:
        //静态的变换只需要发布一次，因为ROS2会为订阅者保留数据，当出现新的订阅者的时候，可以直接获取到保留的数据
        std::shared_ptr<tf2_ros::StaticTransformBroadcaster> broadcaster_;
        void publish_static_tf()
        {
            geometry_msgs::msg::TransformStamped transform;
            transform.header.stamp = this->get_clock()->now();
            transform.header.frame_id = "map";
            transform.child_frame_id = "target_point";
            transform.transform.translation.x = 5.0;
            transform.transform.translation.y = 3.0;
            transform.transform.translation.z = 0.0;
            tf2::Quaternion q;
            q.setRPY(0,0,60/180.0*M_PI);
            transform.transform.rotation.x = q.getX();
            transform.transform.rotation.y = q.getY();
            transform.transform.rotation.z = q.getZ();
            transform.transform.rotation.w = q.getW();
            broadcaster_->sendTransform(transform);
        }
};

int main(int argc,char** argv)
{
    rclcpp::init(argc,argv);
    auto node = std::make_shared<StaticTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: products/cpp/tf_pratice_ws/src/demo_python_tf/demo_python_tf/static_tf_broadcaster.py
import math
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped
from tf2_ros import StaticTransformBroadcaster
from tf_transformations import quaternion_from_euler

class StaticTFBroadcaster(Node):
    def __init__(self,nodename):
        super().__init__(nodename)
        self.static_broadcaster = StaticTransformBroadcaster(self)
        #self.timer = self.create_timer(1,self.broadcast_timer_callback)
    
    def publish_static_tf(self,):
        transform = TransformStamped()
        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "base_link"
        transform.child_frame_id = "camera_link"
        transform.transform.translation.x = 0.5
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.6
        #欧拉角转换四元数
        q = quaternion_from_euler(math.radians(0),0,0)
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]
        #发布静态坐标转换，静态坐标变换只需要发布一次，ROS2会为订阅者保留数据，当出现新的订阅者时，可以直接获取到保留的数据
        self.static_broadcaster.sendTransform(transform)
        self.get_logger().info(f'发布静态坐标转换')

def main():
    rclpy.init()
    node = StaticTFBroadcaster("static_tf_broadcaster")
    node.publish_static_tf()
    rclpy.spin(node)
    rclpy.shutdown()

